src/Microsoft.Health.Checkpoints/Storage/StorageCheckpointClient.cs

```diff
@@ -0,0 +1,143 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+using EnsureThat;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Health.Checkpoints.Storage
+{
+    public class StorageCheckpointClient : ICheckpointClient
+    {
+        private BlobContainerClient _storageClient;
+        private ILogger<StorageCheckpointClient> _logger;
+        private const string _lastProcessedDateTime = "LastProcessedDateTime";
+        private const string _lastProcessedIdentifier = "LastProcessedIdentifier";
+
+        public StorageCheckpointClient(BlobContainerClient containerClient, ILogger<StorageCheckpointClient> logger)
+        {
+            EnsureArg.IsNotNull(containerClient);
+
+            _storageClient = containerClient;
+            _logger = logger;
+        }
+
+        public virtual async Task<ICheckpoint> GetCheckpointAsync(string partition, string checkpointIdentifier)
+        {
+            EnsureArg.IsNotNullOrEmpty(partition, nameof(partition));
+            EnsureArg.IsNotNullOrEmpty(checkpointIdentifier, nameof(checkpointIdentifier));
+
+            ICheckpoint checkpoint = new Checkpoint();
+
+            try
+            {
+                var resultSegment = _storageClient.GetBlobsAsync(
+                                        traits: BlobTraits.Metadata,
+                                        states: BlobStates.All,
+                                        prefix: $"{partition}/{checkpointIdentifier}",
+                                        cancellationToken: CancellationToken.None)
+                                        .AsPages();
+
+                await foreach (Page<BlobItem> blobPage in resultSegment)
+                {
+                    if (blobPage.Values.Count == 0)
+                    {
+                        _logger.LogWarning($"No blob found for identifier {checkpointIdentifier}");
+                    }
+
+                    if (blobPage.Values.Count > 1)
```

Comment:
Does this mean that the last one processed wins? Is that the behavior that we want? Also, looking at the implementation, it seems like we could get a combination of many different checkpoints if the `TryParse` commands below fail.

Reply:
In practice there will only be one blob per partition with checkpoint information.

The blob api makes use specify a "prefix" when searching for a blob. In our case the prefix is just the full blob path so we would only expect 0 or 1 result. In the result, there will be either a lastProcessedDateTime or lastProcessedIdentifier (or both). 

If there is more than 1 result, then that is not expected and we throw an exception.